src/Sarif.Viewer.VisualStudio/Fixes/EditActionPreviewProvider.cs

```diff
@@ -0,0 +1,238 @@
+        private readonly IEditorOptionsFactoryService editorOptionsFactoryService;
+        private readonly IDifferenceBufferFactoryService differenceBufferFactoryService;
+        private readonly IWpfDifferenceViewerFactoryService wpfDifferenceViewerFactoryService;
+
+        [ImportingConstructor]
+        public EditActionPreviewProvider(
+            ITextBufferFactoryService textBufferFactoryService,
+            ITextDocumentFactoryService textDocumentFactoryService,
+            ITextDifferencingSelectorService textDifferencingSelectorService,
+            IProjectionBufferFactoryService projectionBufferFactoryService,
+            IEditorOptionsFactoryService editorOptionsFactoryService,
+            IDifferenceBufferFactoryService differenceBufferFactoryService,
+            IWpfDifferenceViewerFactoryService wpfDifferenceViewerFactoryService,
+            ITextEditorFactoryService textEditorFactoryService)
+        {
+            this.textBufferFactoryService = textBufferFactoryService;
+            this.textDocumentFactoryService = textDocumentFactoryService;
+            this.textDifferencingSelectorService = textDifferencingSelectorService;
+            this.projectionBufferFactoryService = projectionBufferFactoryService;
+            this.editorOptionsFactoryService = editorOptionsFactoryService;
+            this.differenceBufferFactoryService = differenceBufferFactoryService;
+            this.wpfDifferenceViewerFactoryService = wpfDifferenceViewerFactoryService;
+
+            this.previewRoleSet = textEditorFactoryService.CreateTextViewRoleSet(PredefinedTextViewRoles.Analyzable);
+        }
+
+        public Task<object> CreateChangePreviewAsync(
+            ITextBuffer buffer,
+            Action<ITextBuffer,
+            ITextSnapshot> applyEdit,
+            string description = null,
+            FrameworkElement additionalContent = null)
+        {
+            var bufferClone = this.CloneBuffer(buffer);
+
+            applyEdit(bufferClone, buffer.CurrentSnapshot);
+
+            var diffResult = this.ComputeDifferences(buffer, bufferClone);
+            var originalSpans = this.GetOriginalSpans(diffResult);
+            var changedSpans = this.GetChangedSpans(diffResult);
+
+            var originalLineSpans = this.CreateLineSpans(buffer.CurrentSnapshot, originalSpans);
+            var changedLineSpans = this.CreateLineSpans(bufferClone.CurrentSnapshot, changedSpans);
+
+            if (!originalLineSpans.Any())
+            {
+                originalLineSpans = changedLineSpans;
+            }
+
+            if (!(originalLineSpans.Any() && changedLineSpans.Any()))
+            {
+                return Task.FromResult<object>(null);
+            }
+
+            var originalProjectionBuffer = this.projectionBufferFactoryService.CreateProjectionBufferWithoutIndentation(
+                this.editorOptionsFactoryService.GlobalOptions,
+                buffer.CurrentSnapshot,
+                "...",
```

Comment:
>"..." [](start = 16, length = 5)

We may have borrowed the code, but this really should be in the RESX file. (Same below)

Reply:
Well, this one is just a horizontal ellipsis (and I will change it to a Unicode HORIZONTAL ELLIPSIS U+2026) -- I don't see this as localizable.

---
In reply to: [516034698](https://github.com/microsoft/sarif-visualstudio-extension/pull/239#discussion_r516034698) [](ancestors = 516034698)

Reply:
Also DRY'd it out into a constant:
```
            const string Separator = "\u2026"; // Horizontal ellipsis.
```

---
In reply to: [516093993](https://github.com/microsoft/sarif-visualstudio-extension/pull/239#discussion_r516093993) [](ancestors = 516093993,516034698)